### ISSUE TITLE
Insert arbitrary character

### ISSUE THREAD

**Issue by Thom1729** [Hammerspoon/hammerspoon#2130]
In an eventtap handler, I'd like to insert a character that isn't in the current keyboard layout. Can this be accomplished via `keyStroke`? If not, is there a sane way to do it?

**Comment by latenitefilms** [comment]
Yep, sure can. For example:

```lua
hs.eventtap.keyStroke({}, "§")
```

**Comment by Thom1729** [comment]
That only works because § is in the current keyboard layout. If I run `eventtap.keyStroke({}, "❧")`, I get an error because ❧ is not in the current keyboard layout. I don't believe that there is any way to do this with `keyStroke`, or to explicitly create a keyboard event with `eventtap.event` that will insert an arbitrary character.

However, `eventtap.keyStrokes` does the trick. Apparently, [the implementation](https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/eventtap/internal.m#L82) uses [CGEventKeyboardSetUnicodeString](https://developer.apple.com/documentation/coregraphics/1456028-cgeventkeyboardsetunicodestring?language=objc), which does work for arbitrary characters. It's a shame that there doesn't seem to be a way to explicitly create such an event (or series of events) with the core libs, because it would be nice to return it from an eventtap handler. Adding a `eventtap.event.setUnicodeString()` method would resolve this.

**Comment by cmsj** [comment]
I don't have time to test it right now, but I just threw up a branch on my repo which ought to implement both a `:setUnicodeString()` and a `:getUnicodeString()`.

@Thom1729 if you have the wherewithall to build & try, that would be awesome, but don't worry if not (HS is quite non-trivial to build), I'll get to it in the next few days.

**Comment by interpeix** [comment]
That could be super useful, I have a nearly the same problem. I need to remap alt+escape to backslash but i ended always with this error:

> 2019-08-19 23:56:06: 23:56:06 ** Warning:hs.keycode: key '\\' not found in active keymap; using ANSI-standard US keyboard layout as fallback, returning '42'

Here is my lua code:
```lua
local function keyCode(key, modifiers)
   modifiers = modifiers or {}
   return function()
      hs.eventtap.keyStroke(modifiers, key)
   end
end
 
 local function remapKey(modifiers, key, keyCode)
    hs.hotkey.bind(modifiers, key, keyCode, nil, keyCode)
 end
 
 remapKey({'alt'}, 'escape', keyCode("\\"))
````
Does your fix help in my case?

**Comment by interpeix** [comment]
> That could be super useful, I have a nearly the same problem. I need to remap alt+escape to backslash but i ended always with this error:
> 
> > 2019-08-19 23:56:06: 23:56:06 ** Warning:hs.keycode: key '\' not found in active keymap; using ANSI-standard US keyboard layout as fallback, returning '42'
> 
> Here is my lua code:
> 
> ```lua
> local function keyCode(key, modifiers)
>    modifiers = modifiers or {}
>    return function()
>       hs.eventtap.keyStroke(modifiers, key)
>    end
> end
>  
>  local function remapKey(modifiers, key, keyCode)
>     hs.hotkey.bind(modifiers, key, keyCode, nil, keyCode)
>  end
>  
>  remapKey({'alt'}, 'escape', keyCode("\\"))
> ```
> 
> Does your fix help in my case?

I made a workaround using the clipboard. 
```lua
backslash = "\\"
a = "ª"
o = "º"
-- Hotkey backslash text
hs.hotkey.bind({"alt"}, "escape", function ()
  --hs.alert.show("Remove this message after debugging!")
  --hs.eventtap.keyStrokes(jira_text)#don't do this!
  hs.pasteboard.writeObjects(backslash)
  hs.eventtap.keyStroke("cmd", "v")
  end)
--
hs.hotkey.bind({"ctrl", "alt"}, "escape", function()
   hs.pasteboard.writeObjects(a)
   hs.eventtap.keyStroke("cmd", "v")
end)

hs.hotkey.bind({"shift", "alt"}, "escape", function()
   hs.pasteboard.writeObjects(o)
   hs.eventtap.keyStroke("cmd", "v")
end)
```

**Comment by cmsj** [comment]
@interpeix this probably would help with that, but as always, for simple key remapping like you're doing, I would strongly encourage you to use Karabiner Elements. It operates inside the macOS kernel, so it's happening a *lot* earlier in the input processing chain than Hammerspoon is able to (ie it'll be faster and more reliable).

**Comment by cmsj** [comment]
Ok, I was able to test d815b629c030b3fdf63b2fb6fd472c45c23d8c47 and with a few little fixes in 4453e732b72f9fb061d059fe8eb26bff98bf2f81 it works correctly, and will be in the next release. Thanks all!

**Comment by Thom1729** [comment]
It works beautifully. Thank you!
